ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs

```diff
@@ -0,0 +1,250 @@
+        match node with
+        | :? ITopBinding
+        | :? IMemberDeclaration
+        | :? IConstructorDeclaration -> true
+        | _ -> false
+
+    override x.VisitNode(node, context) =
+        if settings.LocalBindings = PushToHintMode.Never && isTopLevelMember node then () else
+
+        for child in node.Children() do
+            if settings.TopLevelMembers = PushToHintMode.Never &&
+               isTopLevelMember child then x.VisitNode(child, context) else
+
+            match child with
+            | :? IFSharpTreeNode as treeNode -> treeNode.Accept(x, context)
+            | _ -> ()
+
+    override x.VisitLambdaExpr(lambda, context) =
+        let result = collectTypeHintAnchorsForLambda lambda
+        context.LocalNodes.AddRange(result)
+
+        x.VisitNode(lambda, context)
+
+    override x.VisitLocalBinding(binding, context) =
+        let result = collectTypeHintAnchorsForBinding binding
+        context.LocalNodes.AddRange(result)
+
+        x.VisitNode(binding, context)
+
+    override x.VisitTopBinding(binding, context) =
+        let result = collectTypeHintAnchorsForBinding binding
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(binding, context)
+
+    override x.VisitMemberDeclaration(memberDecl, context) =
+        let result = collectTypeHintsAnchorsForMember memberDecl
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(memberDecl, context)
+
+    override x.VisitPrimaryConstructorDeclaration(constructor, context) =
+        let result = collectTypeHintAnchorsForConstructor constructor
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(constructor, context)
+
+    override x.VisitSecondaryConstructorDeclaration(constructor, context) =
+        let result = collectTypeHintAnchorsForConstructor constructor
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(constructor, context)
+
+    override x.VisitForEachExpr(forEachExpr, context) =
+        let result = collectTypeHintAnchorsForEachExpr forEachExpr
+        context.LocalNodes.AddRange(result)
+
+        x.VisitNode(forEachExpr, context)
+
+type private PatternsHighlightingProcess(logger: ILogger, fsFile, settingsStore: IContextBoundSettingsStore,
+                                           daemonProcess: IDaemonProcess, settings) =
+    inherit FSharpDaemonStageProcessBase(fsFile, daemonProcess)
+
+    let createTypeHintHighlighting (fcsType: FSharpType) (displayContext: FSharpDisplayContext) range pushToHintMode =
+        let displayContext = displayContext.WithShortTypeNames(true)
+        TypeHintHighlighting(fcsType.Format(displayContext), range, pushToHintMode)
+
+    let getReturnTypeHint (decl: IParameterOwnerMemberDeclaration) pushToHintMode =
+        match decl with
+        | :? IConstructorDeclaration
+        | :? IAccessorDeclaration -> ValueNone
+        | _ ->
+
+        let equalsToken = decl.EqualsToken
+        let patterns = decl.ParameterPatterns
```

Comment:
Do we really need to enumerate all the patterns and store them in a collection? Can we just check that parameter declaration sequence is not empty?